publisher/publisher.go

```diff
@@ -0,0 +1,288 @@
+	if len(logDesc.ID) != 44 {
+		return fmt.Errorf("Invalid log ID length [%d]", len(logDesc.ID))
+	}
+
+	return nil
+}
+
+// CTConfig defines the JSON configuration file schema
+type CTConfig struct {
+	Logs              []LogDescription `json:"logs"`
+	SubmissionRetries int              `json:"submissionRetries"`
+	// This should use the same method as the DNS resolver
+	SubmissionBackoffString string        `json:"submissionBackoff"`
+	SubmissionBackoff       time.Duration `json:"-"`
+
+	BundleFilename string   `json:"intermediateBundleFilename"`
+	IssuerBundle   []string `json:"-"`
+}
+
+type ctSubmissionRequest struct {
+	Chain []string `json:"chain"`
+}
+
+const (
+	sctVersion       = 0
+	sctSigType       = 0
+	sctX509EntryType = 0
+)
+
+// PublisherAuthorityImpl defines a Publisher
+type PublisherAuthorityImpl struct {
+	log *blog.AuditLogger
+	CT  *CTConfig
+	SA  core.StorageAuthority
+}
+
+// NewPublisherAuthorityImpl creates a Publisher that will submit certificates
+// to any CT logs configured in CTConfig
+func NewPublisherAuthorityImpl(ctConfig *CTConfig) (pub PublisherAuthorityImpl, err error) {
+	logger := blog.GetAuditLogger()
+	logger.Notice("Publisher Authority Starting")
+	pub.log = logger
+
+	if ctConfig == nil {
+		err = fmt.Errorf("No CT configuration provided")
+		return
+	}
+	if ctConfig.BundleFilename == "" {
+		err = fmt.Errorf("No CT submission bundle provided")
+		return
+	}
+	pub.CT = ctConfig
+	bundle, err := core.LoadCertBundle(ctConfig.BundleFilename)
+	if err != nil {
+		return
+	}
+	for _, cert := range bundle {
+		pub.CT.IssuerBundle = append(pub.CT.IssuerBundle, base64.StdEncoding.EncodeToString(cert.Raw))
+	}
+	ctBackoff, err := time.ParseDuration(ctConfig.SubmissionBackoffString)
+	if err != nil {
+		return
+	}
+	pub.CT.SubmissionBackoff = ctBackoff
+
+	return
+}
+
+func (pub *PublisherAuthorityImpl) submitToCTLog(serial string, jsonSubmission []byte, log LogDescription, client http.Client) error {
+	done := false
+	var retries int
+	var sct core.SignedCertificateTimestamp
+	for !done && retries <= pub.CT.SubmissionRetries {
+		resp, err := postJSON(&client, fmt.Sprintf("%s%s", log.URI, "/ct/v1/add-chain"), jsonSubmission, &sct)
+		if err != nil {
+			// Retry the request, log the error
+			// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+			pub.log.AuditErr(fmt.Errorf("Error POSTing JSON to CT log submission endpoint [%s]: %s", log.URI, err))
+			if retries >= pub.CT.SubmissionRetries {
+				break
+			}
+			retries++
+			time.Sleep(pub.CT.SubmissionBackoff)
+			continue
+		} else {
+			if resp.StatusCode == http.StatusRequestTimeout || resp.StatusCode == http.StatusServiceUnavailable {
+				// Retry the request after either 10 seconds or the period specified
+				// by the Retry-After header
+				backoff := pub.CT.SubmissionBackoff
+				if retryAfter := resp.Header.Get("Retry-After"); retryAfter != "" {
+					if seconds, err := strconv.Atoi(retryAfter); err != nil {
```

Comment:
You can un-nest this `if`. If the Retry-After header is empty, you'll just get an error from Atoi, right?